patroni/quorum.py

```diff
@@ -0,0 +1,186 @@
+        self.check_invariants()
+        logger.debug('sync %s %s', self.numsync, self.sync)
+        return 'sync', self.numsync, self.sync
+
+    def __iter__(self):
+        transitions = list(self._generate_transitions())
+        # Merge 2 transitions of the same type to a single one. This is always safe because skipping the first
+        # transition is equivalent to no one observing the intermediate state.
+        for cur_transition, next_transition in zip(transitions, transitions[1:]+[None]):
+            if next_transition and cur_transition[0] == next_transition[0]:
+                continue
+            yield cur_transition
+
+    def _generate_transitions(self):
+        logger.debug("Quorum state: quorum %s, voters %s, numsync %s, sync %s, active %s, sync_wanted %s",
+                     self.quorum, self.voters, self.numsync, self.sync, self.active, self.sync_wanted)
+        self.check_invariants()
+
+        # Handle non steady state cases
+        if self.sync < self.voters:
+            logger.debug("Case 1: synchronous_standby_names subset of DCS state")
+            # Case 1: quorum is superset of sync nodes. In the middle of changing quorum.
+            # Evict from quorum dead nodes that are not being synced.
+            remove_from_quorum = self.voters - (self.sync | self.active)
+            if remove_from_quorum:
+                yield self.quorum_update(
+                    quorum=len(self.voters) - len(remove_from_quorum) + 1 - self.numsync,
+                    voters=self.voters - remove_from_quorum)
+            # Start syncing to nodes that are in quorum and alive
+            add_to_sync = self.voters - self.sync
+            if add_to_sync:
+                yield self.sync_update(self.numsync, self.sync | add_to_sync)
+        elif self.sync > self.voters:
+            logger.debug("Case 2: synchronous_standby_names superset of DCS state")
+            # Case 2: sync is superset of quorum nodes. In the middle of changing replication factor.
+            # Add to quorum voters nodes that are already synced and active
+            add_to_quorum = (self.sync - self.voters) & self.active
+            if add_to_quorum:
+                yield self.quorum_update(
+                        quorum=self.quorum,
+                        voters=self.voters | add_to_quorum)
+            # Remove from sync nodes that are dead
+            remove_from_sync = self.sync - self.voters
+            if remove_from_sync:
+                yield self.sync_update(
+                        numsync=min(self.sync_wanted, len(self.sync) - len(remove_from_sync)),
+                        sync=self.sync - remove_from_sync)
+
+        # After handling these two cases quorum and sync must match.
+        assert self.voters == self.sync
+
+        safety_margin = self.quorum + self.numsync - len(self.voters | self.sync)
+        if safety_margin > 1:
+            logger.debug("Case 3: replication factor is bigger than needed")
```

Comment:
This branch is not covered by tests, so I tried to do it:
```python
QuorumStateResolver(quorum=3, voters=set('abcdef'),
                    numsync=5, sync=set('abcdef'),
                    active=set('abcdef'), sync_wanted=3)
```
and got an exception `QuorumError: Quorum and sync not guaranteed to overlap: nodes 6 >= quorum 3 + sync 3` from `sync_update()`